src/hotspot/share/oops/method.cpp

```diff
@@ -566,16 +566,23 @@ MethodCounters* Method::build_method_counters(Thread* current, Method* m) {
   methodHandle mh(current, m);
   MethodCounters* counters;
   if (current->is_Java_thread()) {
-    // For when TRAPS is JavaThread.
-    counters = MethodCounters::allocate(mh, current->as_Java_thread());
+    Thread* THREAD = current;
+    counters = MethodCounters::allocate(mh, THREAD);
+    if (HAS_PENDING_EXCEPTION) {
+      CLEAR_PENDING_EXCEPTION;  // MethodData above doesn't clear exception
+      CompileBroker::log_metaspace_failure();
+      ClassLoaderDataGraph::set_metaspace_oom(true);
+      return NULL;
```

Comment:
You could factor this out for both cases by testing "counters == NULL".

Reply:
Yes, this is better.